pkg/ovsdb/transact.go

```diff
@@ -0,0 +1,711 @@
+}
+
+func (txn *Transaction) Commit() error {
+	var err error
+
+	/* fetch needed data from database needed to perform the operation */
+	for i, ovsOp := range txn.operations {
+		if err = ovsOpCallbackMap[ovsOp.Op][0](txn, &ovsOp, &txn.response.Result[i]); err != nil {
+			txn.response.Error = err.Error()
+			return err
+		}
+	}
+	_, err = txn.etcdCommit()
+	if err != nil {
+		txn.response.Error = err.Error()
+		return err
+	}
+
+	/* commit actual transactional changes to database */
+	for i, ovsOp := range txn.operations {
+		ovsOpCallbackMap[ovsOp.Op][1](txn, &ovsOp, &txn.response.Result[i])
+		txn.response.Error = err.Error()
+		return err
+	}
+	_, err = txn.etcdCommit()
+	if err != nil {
+		txn.response.Error = err.Error()
+		return err
+	}
+
+	return nil
+}
+
+const DELIM = "/"
+const UUID = "_uuid"
+
+func makePrefix(dbname, table string) string {
+	return dbname + DELIM + table
+}
+
+func makeKey(dbname, table, uuid string) string {
+	return dbname + DELIM + table + DELIM + uuid
+}
+
+func makeValue(row map[string]interface{}) (string, error) {
+	b, err := json.Marshal(row)
+	if err != nil {
+		return "", err
+	}
+	return string(b), nil
+}
+
+func setRowUUID(row map[string]interface{}, uuid string) {
+	row[UUID] = uuid
+}
+
+const (
+	FN_LT = "<"
+	FN_LE = "<="
+	FN_EQ = "=="
+	FN_NE = "!="
+	FN_GE = ">="
+	FN_GT = ">"
+	FN_IN = "includes"
+	FN_EX = "excludes"
+)
+
+func isSupportedFunc(function string) error {
+	if function == FN_EQ || function == FN_NE {
+		return nil
+	}
+	return errors.New("not supported")
+}
+
+func parseCond(cond []interface{}) (string, string, interface{}, error) {
+	if len(cond) != 3 {
+		return "", "", nil, errors.New("expected 3 elements in condition")
+	}
+
+	column, ok := cond[0].(string)
+	if !ok {
+		return "", "", nil, errors.New("column must be of type string")
+	}
+
+	fn, ok := cond[1].(string)
+	if !ok {
+		return "", "", nil, errors.New("function must be of type string")
+	}
+	if err := isSupportedFunc(fn); err != nil {
+		return "", "", nil, err
+	}
+
+	value := cond[2]
+	if !ok {
+		return "", "", nil, errors.New("value must be of type string")
+	}
+	return column, fn, value, nil
+}
+
+func condInteger(act interface{}, fn string, exp interface{}) (bool, error) {
+	actual, ok := act.(int)
+	if !ok {
+		return false, errors.New("failed to convert to integer")
+	}
+	expected, ok := exp.(int)
+	if !ok {
+		return false, errors.New("failed to convert to integer")
+	}
+	if (fn == FN_EQ || fn == FN_IN) && actual == expected {
+		return true, nil
+	}
+	if (fn == FN_NE || fn == FN_EX) && actual != expected {
+		return true, nil
+	}
+	if fn == FN_GT && actual > expected {
+		return true, nil
+	}
+	if fn == FN_GE && actual >= expected {
+		return true, nil
+	}
+	if fn == FN_LT && actual < expected {
+		return true, nil
+	}
+	if fn == FN_LE && actual <= expected {
+		return true, nil
+	}
+	return false, nil
+}
+
+func condReal(act interface{}, fn string, exp interface{}) (bool, error) {
```

Comment:
can we combine condInteger and condReal ?

Reply:
no. as we depend on type conversion so while code looks the same one works on int the other on float64 (this is the compromise when using a strongly typed lang such as go)